src/coreclr/jit/helperexpansion.cpp

```diff
@@ -1553,3 +1602,317 @@ bool Compiler::fgVNBasedIntrinsicExpansionForCall_ReadUtf8(BasicBlock** pBlock,
+        *likelihood    = 100;
+        return (CORINFO_CLASS_HANDLE)likelyClasses[index].handle;
+    }
+#endif
+    *likelihood = likelyClasses[0].likelihood;
+    return (CORINFO_CLASS_HANDLE)likelyClasses[0].handle;
+}
+
+//------------------------------------------------------------------------------
+// fgLateCastExpansionForCall : Expand specific cast helper, see
+//    fgLateCastExpansion's comments.
+//
+// Arguments:
+//    block - Block containing the cast helper to expand
+//    stmt  - Statement containing the cast helper
+//    call  - The cast helper
+//
+// Returns:
+//    True if expanded, false otherwise.
+//
+bool Compiler::fgLateCastExpansionForCall(BasicBlock** pBlock, Statement* stmt, GenTreeCall* call)
+{
+    if (!call->IsHelperCall() || !impIsCastHelperMayHaveProfileData(call->GetHelperNum()))
+    {
+        // Not a cast helper we're interested in
+        return false;
+    }
+
+    if ((call->gtCallMoreFlags & GTF_CALL_M_CAST_CAN_BE_EXPANDED) == 0)
+    {
+        // It's not eligible for expansion (already expanded in importer)
+        // To be removed once we move cast expansion here completely.
+        return false;
+    }
+
+    BasicBlock* block = *pBlock;
+    JITDUMP("Attempting to expand a cast helper call in " FMT_BB "...\n", block->bbNum);
+    DISPTREE(call);
+    JITDUMP("\n");
+
+    // Currently, we only expand "isinst" and only using profile data. The long-term plan is to
+    // move cast expansion logic here from the importer completely.
+    unsigned             likelihood = 100;
+    CORINFO_CLASS_HANDLE likelyCls  = PickLikelyClass(this, call->gtCastHelperILOffset, &likelihood);
+    if (likelyCls == NO_CLASS_HANDLE)
+    {
+        // TODO: make null significant, so it could mean our object is likely just null
+        JITDUMP("Likely class is null - bail out.\n");
+        return false;
+    }
+
+    // if there is a dominating candidate with >= 50% likelihood, use it
+    const unsigned likelihoodMinThreshold = 50;
+    if (likelihood < likelihoodMinThreshold)
+    {
+        JITDUMP("Likely class likelihood is below %u%% - bail out.\n", likelihoodMinThreshold);
+        return false;
+    }
+
+    // E.g. "call CORINFO_HELP_ISINSTANCEOFCLASS(class, obj)"
+    GenTree*             clsArg      = call->gtArgs.GetUserArgByIndex(0)->GetNode();
+    CORINFO_CLASS_HANDLE expectedCls = gtGetHelperArgClassHandle(clsArg);
+    if (expectedCls == NO_CLASS_HANDLE)
+    {
+        // clsArg doesn't represent a class handle - bail out
+        // TODO-InlineCast: use VN if it's available (depends on when this phase is executed)
+        JITDUMP("clsArg is not a constant handle - bail out.\n");
+        return false;
+    }
+
+    const TypeCompareState castResult = info.compCompHnd->compareTypesForCast(likelyCls, expectedCls);
+    if (castResult == TypeCompareState::May)
+    {
+        JITDUMP("compareTypesForCast returned May for this candidate\n");
+        return false;
```

Comment:
What's the reason for this check? Is it reachable? (Seems like it's related to shared generics, but why shouldn't we expand in those cases?)

Reply:
it's copied from the existing code in importer. Honestly I don't remember the rationale behind it, let's keep it for now as extra safety and then I'll inspect the diffs it causes.